File: apps/api/src/lib/network-pool.ts
/**
 * Network Pool Manager
 * 
 * Manages Docker networks to prevent exhaustion (EC-8) and ensure proper cleanup (EC-10).
 * Reuses networks instead of creating per-job to avoid "address pools exhausted" errors.
 * 
 * Each network is isolated (--internal) to prevent external egress.
 */

import { spawn } from 'child_process';
import { Mutex } from 'async-mutex';

// ============ Constants ============

const NETWORK_POOL_SIZE = 50;        // Max networks in pool
const NETWORK_PREFIX = 'grader_net_pool';
const ACQUIRE_TIMEOUT_MS = 10000;    // 10s timeout to acquire network
const CLEANUP_INTERVAL_MS = 5 * 60 * 1000; // 5 minutes

// ============ State ============

const networkPool: string[] = [];
const networkInUse = new Set<string>();
const networkMutex = new Mutex();

// ============ Docker Helpers ============

async function dockerExec(args: string[], timeoutMs: number): Promise<{ stdout: string; stderr: string }> {
    return new Promise((resolve, reject) => {
        const proc = spawn('docker', args, { stdio: ['ignore', 'pipe', 'pipe'] });
        let stdout = '';
        let stderr = '';

        const timer = setTimeout(() => {
            proc.kill('SIGKILL');
            reject(new Error(`Docker command timeout: docker ${args.join(' ')}`));
        }, timeoutMs);

        proc.stdout.on('data', (d) => (stdout += d.toString()));
        proc.stderr.on('data', (d) => (stderr += d.toString()));

        proc.on('close', (code) => {
            clearTimeout(timer);
            if (code === 0) return resolve({ stdout, stderr });
            reject(new Error(`Docker command failed (${code}): docker ${args.join(' ')}\n${stdout}\n${stderr}`));
        });

        proc.on('error', (err) => {
            clearTimeout(timer);
            reject(err);
        });
    });
}

// ============ Network Pool Functions ============

/**
 * Acquire a network from the pool (EC-8)
 * Creates new network if pool isn't full, otherwise waits for one to become available
 */
export async function acquireNetwork(): Promise<string> {
    return await networkMutex.runExclusive(async () => {
        // Try to reuse an existing network
        for (const network of networkPool) {
            if (!networkInUse.has(network)) {
                networkInUse.add(network);
                console.log(`[Network] Acquired existing network: ${network}`);
                return network;
            }
        }
        
        // Create new if under limit
        if (networkPool.length < NETWORK_POOL_SIZE) {
            const name = `${NETWORK_PREFIX}_${networkPool.length}`;
            try {
                await dockerExec(['network', 'create', '--internal', name], 8000);
                networkPool.push(name);
                networkInUse.add(name);
                console.log(`[Network] Created new network: ${name}`);
                return name;
            } catch (error) {
                // Network might already exist from previous run
                if (String(error).includes('already exists')) {
                    networkPool.push(name);
                    networkInUse.add(name);
                    console.log(`[Network] Reusing existing network: ${name}`);
                    return name;
                }
                throw error;
            }
        }
        
        // EC-8: Pool exhausted, throw error (caller should retry)
        throw new Error('Network pool exhausted. All networks are in use.');
    });
}

/**
 * Acquire network with retry (EC-8)
 */
export async function acquireNetworkWithRetry(maxRetries = 5, retryDelayMs = 1000): Promise<string> {
    let lastError: Error | null = null;
    
    for (let attempt = 0; attempt < maxRetries; attempt++) {
        try {
            return await acquireNetwork();
        } catch (error) {
            lastError = error as Error;
            if (attempt < maxRetries - 1) {
                console.log(`[Network] Acquire failed, retrying in ${retryDelayMs}ms (attempt ${attempt + 1}/${maxRetries})`);
                await new Promise(r => setTimeout(r, retryDelayMs));
            }
        }
    }
    
    throw lastError || new Error('Network acquire failed after retries');
}

/**
 * Release network back to pool (EC-10)
 * Cleans up any containers still attached to the network
 */
export async function releaseNetwork(name: string): Promise<void> {
    return await networkMutex.runExclusive(async () => {
        if (!networkInUse.has(name)) {
            console.warn(`[Network] Network ${name} not in use, ignoring release`);
            return;
        }
        
        // EC-10: Clean any containers still attached
        try {
            const { stdout } = await dockerExec(
                ['network', 'inspect', name, '-f', '{{range .Containers}}{{.Name}} {{end}}'],
                5000
            );
            
            const containers = stdout.trim().split(' ').filter(Boolean);
            for (const containerName of containers) {
                console.log(`[Network] Disconnecting orphan container ${containerName} from ${name}`);
                await dockerExec(['network', 'disconnect', '-f', name, containerName], 5000).catch(() => {});
            }
        } catch {
            // Network might not exist or other error, that's okay
        }
        
        networkInUse.delete(name);
        console.log(`[Network] Released network: ${name}`);
    });
}

/**
 * Cleanup orphaned networks (EC-10)
 * Removes networks that are in use but have no containers
 */
export async function cleanupOrphanedNetworks(): Promise<number> {
    let cleanedCount = 0;
    
    try {
        // Find all grader networks
        const { stdout } = await dockerExec(
            ['network', 'ls', '--filter', `name=${NETWORK_PREFIX}`, '--format', '{{.Name}}'],
            10000
        );
        
        const existingNetworks = stdout.trim().split('\n').filter(Boolean);
        
        for (const network of existingNetworks) {
            // Check if network has any containers
            try {
                const { stdout: inspect } = await dockerExec(
                    ['network', 'inspect', network, '-f', '{{len .Containers}}'],
                    5000
                );
                
                const containerCount = parseInt(inspect.trim(), 10);
                
                // If network has no containers and is not in our pool, remove it
                if (containerCount === 0 && !networkPool.includes(network)) {
                    console.log(`[Network] Removing orphan network: ${network}`);
                    await dockerExec(['network', 'rm', network], 5000).catch(() => {});
                    cleanedCount++;
                }
            } catch {
                // Network might have been removed already
            }
        }
    } catch (error) {
        console.warn('[Network] Cleanup error:', error);
    }
    
    if (cleanedCount > 0) {
        console.log(`[Network] Cleaned up ${cleanedCount} orphan networks`);
    }
    
    return cleanedCount;
}

/**
 * Get network pool status
 */
export function getNetworkPoolStatus(): {
    total: number;
    available: number;
    inUse: number;
} {
    return {
        total: networkPool.length,
        available: networkPool.length - networkInUse.size,
        inUse: networkInUse.size,
    };
}

/**
 * Initialize network pool (called on server startup)
 */
export async function initializeNetworkPool(): Promise<void> {
    console.log('[Network] Initializing network pool...');
    
    // Cleanup orphaned networks from previous runs
    await cleanupOrphanedNetworks();
    
    // Pre-create a few networks
    const preCreateCount = 5;
    for (let i = 0; i < preCreateCount; i++) {
        const name = `${NETWORK_PREFIX}_${i}`;
        try {
            await dockerExec(['network', 'create', '--internal', name], 8000);
            networkPool.push(name);
            console.log(`[Network] Pre-created network: ${name}`);
        } catch (error) {
            if (String(error).includes('already exists')) {
                networkPool.push(name);
                console.log(`[Network] Reusing existing network: ${name}`);
            }
        }
    }
    
    // Start periodic cleanup
    setInterval(() => {
        cleanupOrphanedNetworks().catch(err => {
            console.error('[Network] Periodic cleanup failed:', err);
        });
    }, CLEANUP_INTERVAL_MS);
    
    console.log(`[Network] Pool initialized with ${networkPool.length} networks`);
}

/**
 * Drain all networks (cleanup on shutdown)
 */
export async function drainNetworkPool(): Promise<void> {
    console.log('[Network] Draining network pool...');
    
    for (const network of networkPool) {
        try {
            // Disconnect all containers first
            const { stdout } = await dockerExec(
                ['network', 'inspect', network, '-f', '{{range .Containers}}{{.Name}} {{end}}'],
                5000
            ).catch(() => ({ stdout: '' }));
            
            for (const containerName of stdout.trim().split(' ').filter(Boolean)) {
                await dockerExec(['network', 'disconnect', '-f', network, containerName], 5000).catch(() => {});
            }
            
            // Remove the network
            await dockerExec(['network', 'rm', network], 5000).catch(() => {});
        } catch {
            // Ignore errors during drain
        }
    }
    
    networkPool.length = 0;
    networkInUse.clear();
    
    console.log('[Network] Network pool drained');
}

File: apps/grader-go/internal/pool/network_pool.go
package pool

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"sync"
	"time"

	"github.com/yourorg/exam-platform/apps/grader-go/internal/docker"
)

const (
	networkPrefix     = "grader_net_"
	maxNetworks       = 50
	networkAcquireTimeout = 10 * time.Second
)

// NetworkPool manages isolated Docker networks to prevent exhaustion
type NetworkPool struct {
	available chan string
	inUse     map[string]bool
	mu        sync.RWMutex
	maxSize   int
	logger    *slog.Logger
	closed    bool
}

// NewNetworkPool creates a network pool
func NewNetworkPool(maxSize int) *NetworkPool {
	if maxSize <= 0 {
		maxSize = maxNetworks
	}

	return &NetworkPool{
		available: make(chan string, maxSize),
		inUse:     make(map[string]bool),
		maxSize:   maxSize,
		logger:    slog.Default().With("component", "netpool"),
	}
}

// Acquire gets an available network or creates one
func (p *NetworkPool) Acquire(ctx context.Context) (string, error) {
	p.mu.RLock()
	if p.closed {
		p.mu.RUnlock()
		return "", fmt.Errorf("network pool is closed")
	}
	p.mu.RUnlock()

	// Try to get from pool (non-blocking)
	select {
	case network := <-p.available:
		p.mu.Lock()
		p.inUse[network] = true
		p.mu.Unlock()
		return network, nil
	default:
	}

	// Check if we can create a new one
	p.mu.Lock()
	totalCount := len(p.inUse) + len(p.available)
	if totalCount >= p.maxSize {
		p.mu.Unlock()
		// Wait for one to become available
		select {
		case network := <-p.available:
			p.mu.Lock()
			p.inUse[network] = true
			p.mu.Unlock()
			return network, nil
		case <-time.After(networkAcquireTimeout):
			return "", fmt.Errorf("network acquire timeout: all %d networks in use", p.maxSize)
		case <-ctx.Done():
			return "", ctx.Err()
		}
	}
	p.mu.Unlock()

	// Create a new network
	network, err := p.createNetwork(ctx)
	if err != nil {
		return "", err
	}

	p.mu.Lock()
	p.inUse[network] = true
	p.mu.Unlock()

	return network, nil
}

// Release returns a network to the pool
func (p *NetworkPool) Release(ctx context.Context, network string) error {
	if network == "" {
		return nil
	}

	p.mu.Lock()
	delete(p.inUse, network)
	p.mu.Unlock()

	// Try to return to pool
	select {
	case p.available <- network:
		// Successfully returned
	default:
		// Pool is full, destroy the network
		if err := docker.RemoveNetwork(ctx, network); err != nil {
			p.logger.Warn("failed to remove network", "name", network, "error", err)
		}
	}

	return nil
}

// createNetwork creates a new isolated Docker network
func (p *NetworkPool) createNetwork(ctx context.Context) (string, error) {
	name := fmt.Sprintf("%s%d_%d", networkPrefix, os.Getpid(), time.Now().UnixNano())

	if err := docker.CreateNetwork(ctx, name); err != nil {
		return "", fmt.Errorf("failed to create network %s: %w", name, err)
	}

	p.logger.Info("created network", "name", name)
	return name, nil
}

// Warm pre-creates networks
func (p *NetworkPool) Warm(ctx context.Context, count int) error {
	if count <= 0 {
		count = 5 // default warmup
	}
	if count > p.maxSize {
		count = p.maxSize
	}

	p.mu.RLock()
	currentSize := len(p.inUse) + len(p.available)
	p.mu.RUnlock()

	toCreate := count - currentSize
	if toCreate <= 0 {
		return nil
	}

	p.logger.Info("warming network pool", "count", toCreate)

	for i := 0; i < toCreate; i++ {
		if ctx.Err() != nil {
			return ctx.Err()
		}

		network, err := p.createNetwork(ctx)
		if err != nil {
			p.logger.Warn("warm create failed", "error", err)
			continue
		}

		select {
		case p.available <- network:
		default:
			docker.RemoveNetwork(ctx, network)
		}
	}

	return nil
}

// Size returns total networks (available + in use)
func (p *NetworkPool) Size() int {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return len(p.inUse) + len(p.available)
}

// Available returns number of available networks
func (p *NetworkPool) Available() int {
	return len(p.available)
}

// InUse returns number of networks currently in use
func (p *NetworkPool) InUse() int {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return len(p.inUse)
}

// Close cleans up all networks
func (p *NetworkPool) Close(ctx context.Context) error {
	p.mu.Lock()
	if p.closed {
		p.mu.Unlock()
		return nil
	}
	p.closed = true
	p.mu.Unlock()

	// Drain available
	close(p.available)
	for network := range p.available {
		docker.RemoveNetwork(ctx, network)
	}

	// Clean up in-use (shouldn't happen normally)
	p.mu.Lock()
	for network := range p.inUse {
		docker.RemoveNetwork(ctx, network)
	}
	p.inUse = make(map[string]bool)
	p.mu.Unlock()

	p.logger.Info("network pool closed")
	return nil
}

// CleanupOrphans removes stale networks from previous runs
func (p *NetworkPool) CleanupOrphans(ctx context.Context) error {
	result, err := docker.Exec(ctx, []string{
		"network", "ls",
		"--filter", fmt.Sprintf("name=%s", networkPrefix),
		"--format", "{{.Name}}",
	}, 10*time.Second)
	if err != nil {
		return fmt.Errorf("failed to list networks: %w", err)
	}

	networks := splitLines(result.Stdout)
	for _, network := range networks {
		if network == "" {
			continue
		}
		p.logger.Info("cleaning orphan network", "name", network)
		docker.RemoveNetwork(ctx, network)
	}

	return nil
}

func splitLines(s string) []string {
	var lines []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == '\n' {
			if i > start {
				lines = append(lines, s[start:i])
			}
			start = i + 1
		}
	}
	if start < len(s) {
		lines = append(lines, s[start:])
	}
	return lines
}
